packages/concerto-core/lib/introspect/mapvaluetype.js

```diff
@@ -64,23 +67,64 @@ class MapValueType extends Decorated {
      * @protected
      */
     validate() {
-        const declarations = this.getModelFile().getAllDeclarations();
-
-        const value = declarations.find(decl => decl.name === this.type);
-
-        if (!value?.isConcept?.()           &&
-            !value?.isEnum?.()              &&
-            !value?.isAsset?.()             &&
-            !value?.isEvent?.()             &&
-            !value?.isParticipant?.()       &&
-            !value?.isTransaction?.()       &&
-            !value?.isMapDeclaration?.()    &&
-            !value?.isScalarDeclaration?.() &&
-            !['String', 'Long', 'Integer', 'Double', 'Boolean', 'DateTime'].includes(this.type)) {
-
-            throw new IllegalModelException(
-                `MapPropertyType has invalid Type: ${this.type}`
-            );
+        if (!ModelUtil.isPrimitiveType(this.type)) {
+            let decl = this.parent.getModelFile().getAllDeclarations().find(d => d.name === this.ast.type?.name);
+
+            // All declarations, with the exception of MapDeclarations are valid Values.
+            if(decl.isMapDeclaration?.()) {
+                throw new IllegalModelException(
+                    `MapDeclaration as Map Type Value is not supported: ${this.type}`
+                );
+            }
+        }
+    }
+
+    /**
+     * Sets the Type name for the Map Value
+     *
+     * @param {Object} ast - The AST created by the parser
+     */
+    processType(ast) {
+        let decl;
+        switch(this.ast.$class) {
+        case `${MetaModelNamespace}.ObjectMapValueType`:
+            decl = this.parent.getModelFile().getAllDeclarations().find(d => d.name === this.ast.type.name);
+
+            // ObjectMapValueType must have TypeIdentifier.
+            if (!('type' in ast)) {
+                throw new IllegalModelException(`ObjectMapValueType must contain property 'type', for MapDeclaration named ${this.parent.name}`);
+            }
+
+            // ObjectMapValueType TypeIdentifier must be properly formed.
+            if (!('$class' in ast.type) || !('name' in ast.type)) {
+                throw new IllegalModelException(`ObjectMapValueType type must contain property '$class' and property 'name', for MapDeclaration named ${this.parent.name}`);
+            }
```

Comment:
Great defensive programming. This should also be checked now by the Metamodel validation in #684, so we may find that we have dead code once both PRs are merged.

Reply:
Noted for cleanup